### PR TITLE
add ruby script to migrate units without importing CB content

### PR DESCRIPTION
**Pull request by davidsbailey**
Continues work on https://codedotorg.atlassian.net/browse/PLAT-830. Most of the code in 0bbea892e538b2d75059398dbdc4d8a3a0d58e70 was copied from https://github.com/code-dot-org/code-dot-org/blob/463c5c603b2de5ed33587c6b6238825b129b87d2/bin/curriculum/import_unit_details.rb .

## Testing story

manually verified as follows:
* `valentine` unit overview page appears unchanged (no View Lesson Plan links)
* `ap-csa` unit overview page appears unchanged (same View Lesson Plan links)
* `AlgebraA` unit can be migrated without errors due to upper case letters in the name
* legacy teacher resources in CSF, CSD and CSP courses from previous years are protected, which I verified as follows
```
Dave-MBP:~/src/cdo (migrate-unit-script)$ bundle exec bin/curriculum/migrate_unit.rb -u coursea-2020 -v
loading rails environment...
rails environment loaded in 33 seconds.
found code studio script name coursea-2020 with id 663
bundler: failed to load command: bin/curriculum/migrate_unit.rb (bin/curriculum/migrate_unit.rb)
RuntimeError: unsafe to migrate launched script containing teacher resources
  bin/curriculum/migrate_unit.rb:71:in `block in main'
  ...
```

```
[development] dashboard > unsafe_scripts = Script.all.select {|script| !script.is_migrated && script.teacher_resources.present? && script.launched?}
[development] dashboard > unsafe_scripts.map(&:name)
=> ["coursea-2018", "coursea-2019", "coursea-2020", "courseb-2018", "courseb-2019", "courseb-2020", "coursec-2018", "coursec-2019", "coursec-2020", "coursed-2018", "coursed-2019", "coursed-2020", "coursee-2018", "coursee-2019", "coursee-2020", "coursef-2018", "coursef-2019", "coursef-2020", "csd1-2017", "csd1-2018", "csd1-2019", "csd1-2020", "csd2-2017", "csd2-2018", "csd2-2019", "csd2-2020", "csd3-2017", "csd3-2018", "csd3-2019", "csd3-2020", "csd3-virtual", "csd4-2017", "csd4-2018", "csd4-2019", "csd4-2020", "csd5-2017", "csd5-2018", "csd5-2019", "csd5-2020", "csd6-2017", "csd6-2018", "csd6-2019", "csd6-2020", "csp-create-2017", "csp-create-2018", "csp-create-2019", "csp-explore-2017", "csp-explore-2018", "csp-explore-2019", "csp1-2017", "csp1-2018", "csp1-2019", "csp1-2020", "csp10-2020", "csp2-2017", "csp2-2018", "csp2-2019", "csp2-2020", "csp3-2017", "csp3-2018", "csp3-2019", "csp3-2020", "csp3-research-mxghyt", "csp4-2017", "csp4-2018", "csp4-2019", "csp4-2020", "csp5-2017", "csp5-2018", "csp5-2019", "csp5-2020", "csp6-2020", "csp7-2020", "csp8-2020", "csp9-2020", "express-2018", "express-2019", "express-2020", "pre-express-2020"]
```

## Deployment strategy

waiting until follow-up work steps are done before serializing any large numbers of scripts on levelbuilder. also waiting for https://github.com/code-dot-org/code-dot-org/pull/42937 to stabilize the script json format.

## Follow-up work

* add an `--all` option to migrate all scripts except the `unsafe_scripts` listed above
* safely handle teacher_resources

### CODE REVIEW

**Comment by davidsbailey** [comment]
It turns out I was missing some units in the beta or published state with legacy teacher resources: 
```
[development] dashboard > Script.all.select {|script| !script.is_migrated && script.teacher_resources.present?}.group_by(&:get_published_state).map {|state, scripts| [state, scripts.count]}.to_h
=> {"beta"=>4, "pilot"=>8, "stable"=>78, "preview"=>1}
```
since none of these are in development, that means we pretty much don't want to migrate any unit with legacy teacher resources, so I removed published_state from the safety check.

**Comment by davidsbailey** [comment]
I discovered when trying to migrate `allthethings` that level swapping is not handled correctly. I've added a check which will prevent the following scripts from being migrated:
```
[development] dashboard > Script.all.select{|s| s.script_levels.any?{|sl| sl.levels.count > 1}}.map(&:name)
=> ["andrea-test", "course2", "coursee-2017", "coursef-2017", "csp-pre-survey", "csp1-2017", "csp1-2018", "csp1-2019", "csp2-2017", "csp3-2017", "csp3-research-mxghyt", "csp3-staging", "csp4-2017", "csp5-2017", "csppostap-2018", "express-2017", "new-d", "new-express", "new-f", "peru-2019", "allthethings"]
```

**Comment by dmcavoy** [comment]
Hey Dave - Looking at some of your recent discoveries on migrating some of these scripts, I would love to talk through out loud the plan for migrating scripts. Are we just not migrating those scripts right now until we figure out how to deal with the cases of teacher resources and level swapping?

**Comment by davidsbailey** [comment]
> Hey Dave - Looking at some of your recent discoveries on migrating some of these scripts, I would love to talk through out loud the plan for migrating scripts. Are we just not migrating those scripts right now until we figure out how to deal with the cases of teacher resources and level swapping?

sure, happy to chat. no, I am planning to migrate those scripts. this is just incremental progress toward that goal.

**Review by dmcavoy — APPROVED** [review]
The script looks good for Scripts that do no have level swapping or teacher resources. I would still love to hear what you are thinking about the process for this moving forward.

File: bin/curriculum/migrate_unit.rb
Comment (dmcavoy) [inline bin/curriculum/migrate_unit.rb:65]: Should the script check that the unit is not already migrated before trying to migrate it?

File: lib/cdo/lesson_import_helper.rb
Comment (bethanyaconnor) [inline lib/cdo/lesson_import_helper.rb]: We should probably remove this line before running this for old scripts. I don't think we want to remove the CurriculumReference levels

**Comment by dmcavoy** [comment]
Bethany's comment actually made me have a couple questions:

1. Are we trying to bring over the lesson plans for all of these old scripts or are we just trying to get the data set up like other courses? 

2. Does this script still handle lockable lesson correctly? I don't know if it was updated after we did all the work on lockable lessons.

**Comment by davidsbailey** [comment]
> Bethany's comment actually made me have a couple questions:
> 
> 1. Are we trying to bring over the lesson plans for all of these old scripts or are we just trying to get the data set up like other courses?

No, there are no more lesson plans that we currently plan to bring over from CB.

> 2. Does this script still handle lockable lesson correctly? I don't know if it was updated after we did all the work on lockable lessons.

Good question! I will check.

**Comment by davidsbailey** [comment]
> > 1. Does this script still handle lockable lesson correctly? I don't know if it was updated after we did all the work on lockable lessons.
> 
> Good question! I will check.

@dmcavoy can you please refresh my memory on what work you are talking about that we did on lockable lessons?

File: lib/cdo/lesson_import_helper.rb
Comment (davidsbailey) [reply lib/cdo/lesson_import_helper.rb]: OK, I've removed this line. we don't currently plan to bring over more lesson plans from CB, so this should be ok. if we did bring more lesson plans over, it seems most likely that they'd be for existing scripts, where we wouldn't want to remove the curriculum reference levels anyway.

**Review by bethanyaconnor — APPROVED** [review]
LGTM pending checking lockable lessons!

**Comment by dmcavoy** [comment]
> > > 1. Does this script still handle lockable lesson correctly? I don't know if it was updated after we did all the work on lockable lessons.
> > 
> > 
> > Good question! I will check.
> 
> @dmcavoy can you please refresh my memory on what work you are talking about that we did on lockable lessons?

Previously we had a system where 'lockable = no lesson plan". We separated those two ideas so now you can set `lockable` and `has_lesson_plan` on a lesson separately. This impacted CSP 20-21 and CSP 21-22 mostly because they had lessons which were lockable and had lesson plans. So I think just check that nothing bad happens to one of the CSP 20-21 scripts (not unit 8 tho) would be the best way to ensure this doesn't need updates

**Comment by davidsbailey** [comment]
lockable lessons look good to me.

so that I could test lockable lessons, I temporarily disabled the check for teacher_resources. once I did that, I verified:
1. no visible differences, aside from the teacher resources button disappearing, which is expected since that's what the check is protecting against:

Screen Shot 2021-10-25 at 2 17 49 PM

Screen Shot 2021-10-25 at 2 18 12 PM

2. links to individual levels and lesson plan are unchanged
3. . no change in result for this command:
```
[development] dashboard > puts Script.find_by_name('csp1-2020').lessons.map {|l| [l.absolute_position, l.lockable, l.has_lesson_plan, l.relative_position, l.name].join("\t")}.join("\n")
1	true	false	1	CS Principles Pre-survey
2	false	true	1	Welcome to CSP
3	false	true	2	Representing Information
4	false	true	3	Circle Square Patterns
5	false	true	4	Binary Numbers
6	false	true	5	Overflow and Rounding
7	false	true	6	Representing Text
8	false	true	7	Black and White Images
9	false	true	8	Color Images
10	false	true	9	Lossless Compression
11	false	true	10	Lossy Compression
12	false	true	11	Intellectual Property
13	false	true	12	Project - Digital Information Dilemmas Part 1
14	false	true	13	Project - Digital Information Dilemmas Part 2
15	true	true	14	Assessment Day
16	false	false	15	CS Principles Post-Course Survey
```

**Comment by davidsbailey** [comment]
> The script looks good for Scripts that do no have level swapping or teacher resources. I would still love to hear what you are thinking about the process for this moving forward.

@dmcavoy here is what I am thinking:
- [x] merge this PR
- [x] wait for the stable sort PR to get to levelbuilder. reserialize script_json files on levelbuilder.
- [ ] ~add `--all` option to this ruby script which makes it migrate all the units which there are not currently problems with~ when that PR gets to levelbuilder, use it to migrate all eligible units  (i.e. not migrated, no teacher resources, no level swaps). this will not include any CSF, CSD, or CSP units.
- [ ] fix level swapping. migrate allthethings unit, which uses level swapping. (see https://github.com/code-dot-org/code-dot-org/pull/43139, only test failures were level swapping)
- [ ] migrate remaining units which use level swapping

at this point, I am left only with the scripts which have teacher resources. there could be several steps and I'll write out a separate plan for that part once I get there.

I know i had suggested using a oneoff script to you for a similar situation, where I want to update content on levelbuilder. that would work here too. I think I just went with this approach because we already have a script written to take command line arguments like this. but I could see this also working as a oneoff script or a series of oneoff scripts.  so I wouldn't hold this up as a generalizable example, but this seemed like the shortest path to me.

**Comment by davidsbailey** [comment]
@dmcavoy do you think we should have a bug bash as part of this process? I could also try to break this up so that I migrate any scripts that are in development or beta without a bash, then do some more focused manual testing for those which are in courses with high amounts of live traffic.

**Comment by dmcavoy** [comment]
> @dmcavoy do you think we should have a bug bash as part of this process? I could also try to break this up so that I migrate any scripts that are in development or beta without a bash, then do some more focused manual testing for those which are in courses with high amounts of live traffic.

it seems like for courses that still have some decent traffic it might be good for us to do a bug bash but I could be convinced otherwise since this years courses aren't being impacted.

**Comment by davidsbailey** [comment]
makes sense. I'm moving this list to a [work plan doc](https://docs.google.com/document/d/1aH_YxmhSXsYcDHmDIU4MDtIu2KoqCgxih6nhGd1SBdo/edit#).

**Pull request merged**
